#! /bin/sh -e

# All lines beginning with `# DPATCH:' are a description of the patch.
# DP: emit correct dwarf info for cfa offset and register with -fomit-frame-pointer

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p1 < $0
        #cd ${dir}gcc && autoconf2.59
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p1 < $0
        #rm ${dir}gcc/configure
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0

---
 gcc/config/m68k/m68k.c |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

Index: gcc-4.1/gcc/config/m68k/m68k.c
===================================================================
--- gcc-4.1.orig/gcc/config/m68k/m68k.c
+++ gcc-4.1/gcc/config/m68k/m68k.c
@@ -604,10 +604,12 @@ m68k_output_function_prologue (FILE *str
 
       if (dwarf2out_do_frame ())
 	{
-	  cfa_offset += current_frame.size + 4;
+	  cfa_offset += current_frame.size;
 	  dwarf2out_def_cfa ("", STACK_POINTER_REGNUM, cfa_offset);
 	}
     } /* !frame_pointer_needed */
+  else if (dwarf2out_do_frame ())
+    dwarf2out_def_cfa ("", STACK_POINTER_REGNUM, cfa_offset);
 
   if (current_frame.fpu_mask)
     {
